File: visitor-pattern/shape/shape.ts
// Abstract visitor
interface ShapeVisitor {
  visitCircle(circle: Shape): number;
  visitRectangle(rectangle: Shape): number;
}

// Concrete visitor
class AreaCalculator implements ShapeVisitor {
  visitCircle(circle: Circle): number {
    return Math.PI * Math.pow(circle.radius, 2);
  }

  visitRectangle(rectangle: Rectangle): number {
    return rectangle.width * rectangle.height;
  }
}

// Abstract shape
interface Shape {
  accept(visitor: ShapeVisitor): number;
}

// <------- Concrete shapes -------->
class Circle implements Shape {
  private _radius: number;

  constructor(rad: number) {
    this._radius = rad;
  }

  get radius() {
    return this._radius;
  }

  accept(visitor: ShapeVisitor): number {
    return visitor.visitCircle(this);
  }
}

class Rectangle implements Shape {
  private _width: number;
  private _height: number;

  constructor(height: number, width: number) {
    this._width = width;
    this._height = height;
  }

  get width() {
    return this._width;
  }

  get height() {
    return this._height;
  }

  accept(visitor: ShapeVisitor): number {
    return visitor.visitRectangle(this);
  }
}

// Dry run
const c1 = new Circle(10);
const r1 = new Rectangle(5, 7);

const visitor = new AreaCalculator();
console.log(`Area of circle: ${c1.accept(visitor)}`);
console.log(`Area of rectangle: ${r1.accept(visitor)}`);

File: visitor-pattern/shape/shape.rs
// Visitor trait
trait ShapeVisitor {
    fn visit_circle(&self, circle: &Circle) -> f64;
    fn visit_rectangle(&self, rectangle: &Rectangle) -> f64;
}

// Concrete visitor
struct AreaCalculator;

impl ShapeVisitor for AreaCalculator {
    fn visit_circle(&self, circle: &Circle) -> f64 {
        circle.radius * circle.radius * std::f64::consts::PI
    }

    fn visit_rectangle(&self, rectangle: &Rectangle) -> f64 {
        rectangle.width * rectangle.height
    }
}

struct PerimeterCalculator;

impl ShapeVisitor for PerimeterCalculator {
    fn visit_circle(&self, circle: &Circle) -> f64 {
        2.0 * std::f64::consts::PI * circle.radius
    }

    fn visit_rectangle(&self, rectangle: &Rectangle) -> f64 {
        2.0 * (rectangle.width + rectangle.height)
    }
}

// Shape trait
trait Shape {
    fn accept(&self, visitor: &dyn ShapeVisitor) -> f64;
}

// Concrete shapes
struct Circle {
    radius: f64,
}

impl Shape for Circle {
    fn accept(&self, visitor: &dyn ShapeVisitor) -> f64 {
        visitor.visit_circle(self)
    }
}

struct Rectangle {
    width: f64,
    height: f64,
}

impl Shape for Rectangle {
    fn accept(&self, visitor: &dyn ShapeVisitor) -> f64 {
        visitor.visit_rectangle(self)
    }
}

fn main() {
    let circle = Circle { radius: 5.0 };
    let rectangle = Rectangle {
        width: 4.0,
        height: 6.0,
    };

    let area_calculator = AreaCalculator;

    println!("Circle area: {}", circle.accept(&area_calculator));
    println!("Rectangle area: {}", rectangle.accept(&area_calculator));

    let perimeter_calculator = PerimeterCalculator;

    println!("Circle perimeter: {}", circle.accept(&perimeter_calculator));
    println!(
        "Rectangle perimeter: {}",
        rectangle.accept(&perimeter_calculator)
    );
}
